README.md
# c-periphery [![Build Status](https://travis-ci.org/vsergeev/c-periphery.svg?branch=master)](https://travis-ci.org/vsergeev/c-periphery) [![GitHub release](https://img.shields.io/github/release/vsergeev/c-periphery.svg?maxAge=7200)](https://github.com/vsergeev/c-periphery) [![License](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/vsergeev/c-periphery/blob/master/LICENSE)

## C Library for Linux Peripheral I/O (GPIO, LED, PWM, SPI, I2C, MMIO, Serial)

c-periphery is a small C library for GPIO, LED, PWM, SPI, I2C, MMIO, and Serial peripheral I/O interface access in userspace Linux. c-periphery simplifies and consolidate the native Linux APIs to these interfaces. c-periphery is useful in embedded Linux environments (including Raspberry Pi, BeagleBone, etc. platforms) for interfacing with external peripherals. c-periphery is re-entrant, has no dependencies outside the standard C library and Linux, compiles into a static library for easy integration with other projects, and is MIT licensed.

Using Python or Lua? Check out the [python-periphery](https://github.com/vsergeev/python-periphery) and [lua-periphery](https://github.com/vsergeev/lua-periphery) projects.

## Examples

### GPIO

``` c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#include "gpio.h"

int main(void) {
    gpio_t *gpio_in, *gpio_out;
    bool value;

    gpio_in = gpio_new();
    gpio_out = gpio_new();

    /* Open GPIO /dev/gpiochip0 line 10 with input direction */
    if (gpio_open(gpio_in, "/dev/gpiochip0", 10, GPIO_DIR_IN) < 0) {
        fprintf(stderr, "gpio_open(): %s\n", gpio_errmsg(gpio_in));
        exit(1);
    }

    /* Open GPIO /dev/gpiochip0 line 12 with output direction */
    if (gpio_open(gpio_out, "/dev/gpiochip0", 12, GPIO_DIR_OUT) < 0) {
        fprintf(stderr, "gpio_open(): %s\n", gpio_errmsg(gpio_out));
        exit(1);
    }

    /* Read input GPIO into value */
    if (gpio_read(gpio_in, &value) < 0) {
        fprintf(stderr, "gpio_read(): %s\n", gpio_errmsg(gpio_in));
        exit(1);
    }

    /* Write output GPIO with !value */
    if (gpio_write(gpio_out, !value) < 0) {
        fprintf(stderr, "gpio_write(): %s\n", gpio_errmsg(gpio_out));
        exit(1);
    }

    gpio_close(gpio_in);
    gpio_close(gpio_out);

    gpio_free(gpio_in);
    gpio_free(gpio_out);

    return 0;
}
```

[Go to GPIO documentation.](docs/gpio.md)

### LED

``` c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#include "led.h"

int main(void) {
    led_t *led;
    unsigned int max_brightness;

    led = led_new();

    /* Open LED led0 */
    if (led_open(led, "led0") < 0) {
        fprintf(stderr, "led_open(): %s\n", led_errmsg(led));
        exit(1);
    }

    /* Enable LED */
    if (led_write(led, true) < 0) {
        fprintf(stderr, "led_write(): %s\n", led_errmsg(led));
        exit(1);
    }

    /* Get max brightness */
    if (led_get_max_brightness(led, &max_brightness) < 0) {
        fprintf(stderr, "led_get_max_brightness(): %s\n", led_errmsg(led));
        exit(1);
    }

    /* Set half brightness */
    if (led_set_brightness(led, max_brightness / 2) < 0) {
        fprintf(stderr, "led_set_brightness(): %s\n", led_errmsg(led));
        exit(1);
    }

    led_close(led);

    led_free(led);

    return 0;
}
```

[Go to LED documentation.](docs/led.md)

### PWM

``` c
#include <stdio.h>
#include <stdlib.h>

#include "pwm.h"

int main(void) {
    pwm_t *pwm;

    pwm = pwm_new();

    /* Open PWM chip 0, channel 10 */
    if (pwm_open(pwm, 0, 10) < 0) {
        fprintf(stderr, "pwm_open(): %s\n", pwm_errmsg(pwm));
        exit(1);
    }

    /* Set frequency to 1 kHz */
    if (pwm_set_frequency(pwm, 1e3) < 0) {
        fprintf(stderr, "pwm_set_frequency(): %s\n", pwm_errmsg(pwm));
        exit(1);
    }

    /* Set duty cycle to 75% */
    if (pwm_set_duty_cycle(pwm, 0.75) < 0) {
        fprintf(stderr, "pwm_set_duty_cycle(): %s\n", pwm_errmsg(pwm));
        exit(1);
    }

    /* Enable PWM */
    if (pwm_enable(pwm) < 0) {
        fprintf(stderr, "pwm_enable(): %s\n", pwm_errmsg(pwm));
        exit(1);
    }

    /* Change duty cycle to 50% */
    if (pwm_set_duty_cycle(pwm, 0.50) < 0) {
        fprintf(stderr, "pwm_set_duty_cycle(): %s\n", pwm_errmsg(pwm));
        exit(1);
    }

    pwm_close(pwm);

    pwm_free(pwm);

    return 0;
}
```

[Go to PWM documentation.](docs/pwm.md)

### SPI

``` c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

#include "spi.h"

int main(void) {
    spi_t *spi;
    uint8_t buf[4] = { 0xaa, 0xbb, 0xcc, 0xdd };

    spi = spi_new();

    /* Open spidev1.0 with mode 0 and max speed 1MHz */
    if (spi_open(spi, "/dev/spidev1.0", 0, 1000000) < 0) {
        fprintf(stderr, "spi_open(): %s\n", spi_errmsg(spi));
        exit(1);
    }

    /* Shift out and in 4 bytes */
    if (spi_transfer(spi, buf, buf, sizeof(buf)) < 0) {
        fprintf(stderr, "spi_transfer(): %s\n", spi_errmsg(spi));
        exit(1);
    }

    printf("shifted in: 0x%02x 0x%02x 0x%02x 0x%02x\n", buf[0], buf[1], buf[2], buf[3]);

    spi_close(spi);

    spi_free(spi);

    return 0;
}
```

[Go to SPI documentation.](docs/spi.md)

### I2C

``` c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

#include "i2c.h"

#define EEPROM_I2C_ADDR 0x50

int main(void) {
    i2c_t *i2c;

    i2c = i2c_new();

    /* Open the i2c-0 bus */
    if (i2c_open(i2c, "/dev/i2c-0") < 0) {
        fprintf(stderr, "i2c_open(): %s\n", i2c_errmsg(i2c));
        exit(1);
    }

    /* Read byte at address 0x100 of EEPROM */
    uint8_t msg_addr[2] = { 0x01, 0x00 };
    uint8_t msg_data[1] = { 0xff, };
    struct i2c_msg msgs[2] =
        {
            /* Write 16-bit address */
            { .addr = EEPROM_I2C_ADDR, .flags = 0, .len = 2, .buf = msg_addr },
            /* Read 8-bit data */
            { .addr = EEPROM_I2C_ADDR, .flags = I2C_M_RD, .len = 1, .buf = msg_data},
        };

    /* Transfer a transaction with two I2C messages */
    if (i2c_transfer(i2c, msgs, 2) < 0) {
        fprintf(stderr, "i2c_transfer(): %s\n", i2c_errmsg(i2c));
        exit(1);
    }

    printf("0x%02x%02x: %02x\n", msg_addr[0], msg_addr[1], msg_data[0]);

    i2c_close(i2c);

    i2c_free(i2c);

    return 0;
}
```

[Go to I2C documentation.](docs/i2c.md)

### MMIO

``` c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <byteswap.h>

#include "mmio.h"

struct am335x_rtcss_registers {
    uint32_t seconds;       /* 0x00 */
    uint32_t minutes;       /* 0x04 */
    uint32_t hours;         /* 0x08 */
    /* ... */
};

int main(void) {
    mmio_t *mmio;
    uint32_t mac_id0_lo, mac_id0_hi;
    volatile struct am335x_rtcss_registers *regs;

    mmio = mmio_new();

    /* Open Control Module */
    if (mmio_open(mmio, 0x44E10000, 0x1000) < 0) {
        fprintf(stderr, "mmio_open(): %s\n", mmio_errmsg(mmio));
        exit(1);
    }

    /* Read lower 2 bytes of MAC address */
    if (mmio_read32(mmio, 0x630, &mac_id0_lo) < 0) {
        fprintf(stderr, "mmio_read32(): %s\n", mmio_errmsg(mmio));
        exit(1);
    }

    /* Read upper 4 bytes of MAC address */
    if (mmio_read32(mmio, 0x634, &mac_id0_hi) < 0) {
        fprintf(stderr, "mmio_read32(): %s\n", mmio_errmsg(mmio));
        exit(1);
    }

    printf("MAC address: %08X%04X\n", __bswap_32(mac_id0_hi), __bswap_16(mac_id0_lo));

    mmio_close(mmio);

    /* Open RTC subsystem */
    if (mmio_open(mmio, 0x44E3E000, 0x1000) < 0) {
        fprintf(stderr, "mmio_open(): %s\n", mmio_errmsg(mmio));
        exit(1);
    }

    regs = mmio_ptr(mmio);

    /* Read current RTC time */
    printf("hours: %02x minutes: %02x seconds %02x\n", regs->hours, regs->minutes, regs->seconds);

    mmio_close(mmio);

    mmio_free(mmio);

    return 0;
}
```

[Go to MMIO documentation.](docs/mmio.md)

### Serial

``` c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "serial.h"

int main(void) {
    serial_t *serial;
    uint8_t s[] = "Hello World!";
    uint8_t buf[128];
    int ret;

    serial = serial_new();

    /* Open /dev/ttyUSB0 with baudrate 115200, and defaults of 8N1, no flow control */
    if (serial_open(serial, "/dev/ttyUSB0", 115200) < 0) {
        fprintf(stderr, "serial_open(): %s\n", serial_errmsg(serial));
        exit(1);
    }

    /* Write to the serial port */
    if (serial_write(serial, s, sizeof(s)) < 0) {
        fprintf(stderr, "serial_write(): %s\n", serial_errmsg(serial));
        exit(1);
    }

    /* Read up to buf size or 2000ms timeout */
    if ((ret = serial_read(serial, buf, sizeof(buf), 2000)) < 0) {
        fprintf(stderr, "serial_read(): %s\n", serial_errmsg(serial));
        exit(1);
    }

    printf("read %d bytes: _%s_\n", ret, buf);

    serial_close(serial);

    serial_free(serial);

    return 0;
}
```

[Go to Serial documentation.](docs/serial.md)

## Building

`make` will build c-periphery into a static library.

``` console
$ make
mkdir obj
cc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  -c src/gpio.c -o obj/gpio.o
cc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  -c src/led.c -o obj/led.o
cc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  -c src/pwm.c -o obj/pwm.o
cc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  -c src/spi.c -o obj/spi.o
cc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  -c src/i2c.c -o obj/i2c.o
cc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  -c src/mmio.c -o obj/mmio.o
cc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  -c src/serial.c -o obj/serial.o
cc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  -c src/version.c -o obj/version.o
ar rcs periphery.a obj/gpio.o obj/led.o obj/pwm.o obj/spi.o obj/i2c.o obj/mmio.o obj/serial.o obj/version.o
$
```

`make tests` will build the c-periphery tests.

``` console
$ make tests
cc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  tests/test_led.c periphery.a -o tests/test_led -lpthread
cc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  tests/test_serial.c periphery.a -o tests/test_serial -lpthread
cc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  tests/test_i2c.c periphery.a -o tests/test_i2c -lpthread
cc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  tests/test_gpio_sysfs.c periphery.a -o tests/test_gpio_sysfs -lpthread
cc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  tests/test_mmio.c periphery.a -o tests/test_mmio -lpthread
cc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  tests/test_spi.c periphery.a -o tests/test_spi -lpthread
cc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  tests/test_gpio.c periphery.a -o tests/test_gpio -lpthread
cc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  tests/test_pwm.c periphery.a -o tests/test_pwm -lpthread
$
```

### Cross-compilation

Set the `CROSS_COMPILE` environment variable with the cross-compiler prefix when calling make:

``` console
$ CROSS_COMPILE=arm-linux- make all tests
mkdir obj
arm-linux-gcc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  -c src/gpio.c -o obj/gpio.o
arm-linux-gcc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  -c src/led.c -o obj/led.o
arm-linux-gcc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  -c src/pwm.c -o obj/pwm.o
arm-linux-gcc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  -c src/spi.c -o obj/spi.o
arm-linux-gcc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  -c src/i2c.c -o obj/i2c.o
arm-linux-gcc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  -c src/mmio.c -o obj/mmio.o
arm-linux-gcc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  -c src/serial.c -o obj/serial.o
arm-linux-gcc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  -c src/version.c -o obj/version.o
arm-linux-ar rcs periphery.a obj/gpio.o obj/led.o obj/pwm.o obj/spi.o obj/i2c.o obj/mmio.o obj/serial.o obj/version.o
arm-linux-gcc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  tests/test_led.c periphery.a -o tests/test_led -lpthread
arm-linux-gcc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  tests/test_serial.c periphery.a -o tests/test_serial -lpthread
arm-linux-gcc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  tests/test_i2c.c periphery.a -o tests/test_i2c -lpthread
arm-linux-gcc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  tests/test_gpio_sysfs.c periphery.a -o tests/test_gpio_sysfs -lpthread
arm-linux-gcc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  tests/test_mmio.c periphery.a -o tests/test_mmio -lpthread
arm-linux-gcc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  tests/test_spi.c periphery.a -o tests/test_spi -lpthread
arm-linux-gcc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  tests/test_gpio.c periphery.a -o tests/test_gpio -lpthread
arm-linux-gcc -std=gnu99 -pedantic -Wall -Wextra -Wno-unused-parameter  -fPIC -DPERIPHERY_VERSION_COMMIT=\"v2.1.0\"  tests/test_pwm.c periphery.a -o tests/test_pwm -lpthread
$
```

## Building c-periphery into another project

Include the header files in `src/` (e.g. `gpio.h`, `led.h`, `pwm.h`, `spi.h`, `i2c.h`, `mmio.h`, `serial.h`) and link in the `periphery.a` static library.

``` console
$ gcc -I/path/to/periphery/src myprog.c /path/to/periphery/periphery.a -o myprog
```

## Documentation

`man` page style documentation for each interface wrapper is available in [docs](docs/) folder.

## Testing

The tests located in the [tests](tests/) folder may be run to test the correctness and functionality of c-periphery. Some tests require interactive probing (e.g. with an oscilloscope), the installation of a physical loopback, or the existence of a particular device on a bus. See the usage of each test for more details on the required test setup.

## License

c-periphery is MIT licensed. See the included [LICENSE](LICENSE) file.



ChangeLog.md
* v2.1.0 - FIXME
    * Add LED module.
    * Add PWM module.
    * Improve string handling in SPI and GPIO modules.

* v2.0.1 - 10/08/2019
    * Initialize handle state in new functions of all modules.
    * Fix performance of blocking read in `serial_read()`.
    * Return error on unexpected empty read in `serial_read()`, which may be
      caused by a serial port disconnect.
    * Improve formatting of `spi_tostring()`.
    * Fix typo in GPIO module documentation.
    * Fix cross-compilation support in Makefile to allow override of CC
      variable.

* v2.0.0 - 09/30/2019
    * Add support for character device GPIOs (`gpio-cdev`) to the GPIO module.
        * Remove support for preserve direction in `gpio_open()`.
        * Remove problematic dummy read with sysfs GPIOs from `gpio_poll()`.
        * Unexport sysfs GPIOs in `gpio_close()`.
    * Migrate to opaque handles with new/free functions in all modules.
    * Simplify error codes for MMIO, I2C, and Serial modules.
    * Fix typos in GPIO module documentation.
    * Update tests with running hints for Raspberry Pi 3.
    * Improve cross-compilation support in Makefile.
    * Contributors
        * longsky, @wangqiang1588 - d880ef7
        * jhlim, @johlim - 742d983

* v1.1.3 - 04/28/2018
    * Fix data's most significant bit getting stripped when opening a serial
      port with parity enabled in `serial_open_advanced()`.
    * Contributors
        * Ryan Barnett, @rjbarnet - 537eeac

* v1.1.2 - 04/01/2018
    * Add handling for delayed pin directory export on some platforms in
      `gpio_open()`.
    * Fix supported functions query for 64-bit in `i2c_open()`.
    * Add support for building with C++.
    * Contributors
        * Jared Bents, @jmbents - 304faf4
        * Ryan Barnett, @rjbarnet - 82ebb4f

* v1.1.1 - 04/25/2017
    * Fix blocking `gpio_poll()` for some platforms.
    * Add library version macros and functions.
    * Contributors
        * Михаил Корнилов, @iTiky - 0643fe9

* v1.1.0 - 09/27/2016
    * Add support for preserving pin direction to `gpio_open()`.
    * Fix enabling input parity check in `serial_set_parity()`.
    * Fix enabling hardware flow control in `serial_set_rtscts()`.
    * Include missing header to fix build with musl libc.
    * Omit unsupported serial baudrates to fix build on SPARC.
    * Contributors
        * Thomas Petazzoni - 27a9552, 114c715

* v1.0.3 - 05/25/2015
    * Fix portability of serial baud rate set/get with termios-provided baud rate functions.
    * Fix unlikely bug in `spi_tostring()` formatting.
    * Clean up integer argument signedness in serial API.

* v1.0.2 - 01/31/2015
    * Fix `gpio_supports_interrupts()` so it does not return an error if interrupts are not supported.
    * Fix `errno` preservation in a few error paths, mostly in the open functions.

* v1.0.1 - 12/26/2014
    * Improve Makefile.
    * Fix _BSD_SOURCE compilation warnings.

* v1.0.0 - 05/15/2014
    * Initial release.


tests/test_led.c
/*
 * c-periphery
 * https://github.com/vsergeev/c-periphery
 * License: MIT
 */

#include "test.h"

#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <unistd.h>

#include "../src/led.h"

const char *device;

void test_arguments(void) {
    ptest();

    /* No real argument validation needed in the LED wrapper */
}

void test_open_config_close(void) {
    led_t *led;
    char name[16];
    unsigned int max_brightness;
    unsigned int brightness;
    bool value;

    ptest();

    /* Allocate LED */
    led = led_new();
    passert(led != NULL);

    /* Open non-existent LED */
    passert(led_open(led, "nonexistent") == LED_ERROR_OPEN);

    /* Open legitimate LED */
    passert(led_open(led, device) == 0);

    /* Check properties */
    passert(led_fd(led) > 0);
    passert(led_name(led, name, sizeof(name)) == 0);
    passert(strcmp(name, device) == 0);

    /* Check max brightness */
    passert(led_get_max_brightness(led, &max_brightness) == 0);
    passert(max_brightness > 0);

    /* Check setting invalid brightness */
    passert(led_set_brightness(led, max_brightness + 1) == LED_ERROR_ARG);

    /* Set brightness to 1, check brightness */
    passert(led_set_brightness(led, 1) == 0);
    usleep(10000);
    passert(led_get_brightness(led, &brightness) == 0);
    passert(brightness >= 1);

    /* Set brightness to 0, check brightness */
    passert(led_set_brightness(led, 0) == 0);
    usleep(10000);
    passert(led_get_brightness(led, &brightness) == 0);
    passert(brightness == 0);

    /* Set brightness to 1, check brightness */
    passert(led_set_brightness(led, 1) == 0);
    usleep(10000);
    passert(led_get_brightness(led, &brightness) == 0);
    passert(brightness >= 1);

    /* Set brightness to 0, check brightness */
    passert(led_set_brightness(led, 0) == 0);
    usleep(10000);
    passert(led_get_brightness(led, &brightness) == 0);
    passert(brightness == 0);

    /* Set brightness to true, check brightness */
    passert(led_write(led, true) == 0);
    usleep(10000);
    passert(led_read(led, &value) == 0);
    passert(value == true);
    passert(led_get_brightness(led, &brightness) == 0);
    passert(brightness > 0);

    /* Set brightness to false, check brightness */
    passert(led_write(led, false) == 0);
    usleep(10000);
    passert(led_read(led, &value) == 0);
    passert(value == false);
    passert(led_get_brightness(led, &brightness) == 0);
    passert(brightness == 0);

    passert(led_close(led) == 0);

    /* Free LED */
    led_free(led);
}

void test_loopback(void) {
    ptest();
}

bool getc_yes(void) {
    char buf[4];
    fgets(buf, sizeof(buf), stdin);
    return (buf[0] == 'y' || buf[0] == 'Y');
}

void test_interactive(void) {
    char str[256];
    led_t *led;

    ptest();

    /* Allocate LED */
    led = led_new();
    passert(led != NULL);

    passert(led_open(led, device) == 0);

    printf("Starting interactive test...\n");
    printf("Press enter to continue...\n");
    getc(stdin);

    /* Check tostring */
    passert(led_tostring(led, str, sizeof(str)) > 0);
    printf("LED description: %s\n", str);
    printf("LED description looks OK? y/n\n");
    passert(getc_yes());

    /* Turn LED off */
    passert(led_write(led, false) == 0);
    printf("LED is off? y/n\n");
    passert(getc_yes());

    /* Turn LED on */
    passert(led_write(led, true) == 0);
    printf("LED is on? y/n\n");
    passert(getc_yes());

    /* Turn LED off */
    passert(led_write(led, false) == 0);
    printf("LED is off? y/n\n");
    passert(getc_yes());

    /* Turn LED on */
    passert(led_write(led, true) == 0);
    printf("LED is on? y/n\n");
    passert(getc_yes());

    passert(led_close(led) == 0);

    /* Free LED */
    led_free(led);
}

int main(int argc, char *argv[]) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s <LED name>\n\n", argv[0]);
        fprintf(stderr, "[1/4] Arguments test: No requirements.\n");
        fprintf(stderr, "[2/4] Open/close test: LED should be real.\n");
        fprintf(stderr, "[3/4] Loopback test: No test.\n");
        fprintf(stderr, "[4/4] Interactive test: LED should be observed.\n\n");
        fprintf(stderr, "Hint: for Raspberry Pi 3, disable triggers for led1:\n");
        fprintf(stderr, "    $ echo none > /sys/class/leds/led1/trigger\n");
        fprintf(stderr, "Observe led1 (red power LED), and run this test:\n");
        fprintf(stderr, "    %s led1\n\n", argv[0]);
        exit(1);
    }

    device = argv[1];

    test_arguments();
    printf(" " STR_OK "  Arguments test passed.\n\n");
    test_open_config_close();
    printf(" " STR_OK "  Open/close test passed.\n\n");
    test_loopback();
    printf(" " STR_OK "  Loopback test passed.\n\n");
    test_interactive();
    printf(" " STR_OK "  Interactive test passed.\n\n");

    printf("All tests passed!\n");
    return 0;
}



src/led.h
/*
 * c-periphery
 * https://github.com/vsergeev/c-periphery
 * License: MIT
 */

#ifndef _PERIPHERY_LED_H
#define _PERIPHERY_LED_H

#ifdef __cplusplus
extern "C" {
#endif

#include <stddef.h>
#include <stdbool.h>
#include <stdint.h>

enum led_error_code {
    LED_ERROR_ARG       = -1, /* Invalid arguments */
    LED_ERROR_OPEN      = -2, /* Opening LED */
    LED_ERROR_QUERY     = -3, /* Querying LED attributes */
    LED_ERROR_IO        = -4, /* Reading/writing LED brightness */
    LED_ERROR_CLOSE     = -5, /* Closing LED */
};

typedef struct led_handle led_t;

/* Primary Functions */
led_t *led_new(void);
int led_open(led_t *led, const char *name);
int led_read(led_t *led, bool *value);
int led_write(led_t *led, bool value);
int led_close(led_t *led);
void led_free(led_t *led);

/* Getters */
int led_get_brightness(led_t *led, unsigned int *brightness);
int led_get_max_brightness(led_t *led, unsigned int *max_brightness);

/* Setters */
int led_set_brightness(led_t *led, unsigned int brightness);

/* Miscellaneous */
int led_fd(led_t *led);
int led_name(led_t *led, char *str, size_t len);
int led_tostring(led_t *led, char *str, size_t len);

/* Error Handling */
int led_errno(led_t *led);
const char *led_errmsg(led_t *led);

#ifdef __cplusplus
}
#endif

#endif



src/led.c
/*
 * c-periphery
 * https://github.com/vsergeev/c-periphery
 * License: MIT
 */

#include <stddef.h>
#include <stdbool.h>
#include <stdarg.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include <unistd.h>
#include <fcntl.h>
#include <errno.h>

#include "led.h"

#define P_PATH_MAX  256

struct led_handle {
    int fd;
    char name[64];
    unsigned int max_brightness;

    struct {
        int c_errno;
        char errmsg[96];
    } error;
};

static int _led_error(led_t *led, int code, int c_errno, const char *fmt, ...) {
    va_list ap;

    led->error.c_errno = c_errno;

    va_start(ap, fmt);
    vsnprintf(led->error.errmsg, sizeof(led->error.errmsg), fmt, ap);
    va_end(ap);

    /* Tack on strerror() and errno */
    if (c_errno) {
        char buf[64];
        strerror_r(c_errno, buf, sizeof(buf));
        snprintf(led->error.errmsg+strlen(led->error.errmsg), sizeof(led->error.errmsg)-strlen(led->error.errmsg), ": %s [errno %d]", buf, c_errno);
    }

    return code;
}

led_t *led_new(void) {
    led_t *led = calloc(1, sizeof(led_t));
    if (led == NULL)
        return NULL;

    led->fd = -1;

    return led;
}

int led_open(led_t *led, const char *name) {
    char led_path[P_PATH_MAX];
    int fd, ret;

    snprintf(led_path, sizeof(led_path), "/sys/class/leds/%s/brightness", name);

    if ((fd = open(led_path, O_RDWR)) < 0)
        return _led_error(led, LED_ERROR_OPEN, errno, "Opening LED: opening 'brightness'");

    led->fd = fd;
    strncpy(led->name, name, sizeof(led->name));

    if ((ret = led_get_max_brightness(led, &led->max_brightness)) < 0)
        return ret;

    return 0;
}

int led_read(led_t *led, bool *value) {
    int ret;
    unsigned int brightness;

    if ((ret = led_get_brightness(led, &brightness)) < 0)
        return ret;

    *value = brightness != 0;

    return 0;
}

int led_write(led_t *led, bool value) {
    int ret;

    if ((ret = led_set_brightness(led, value ? led->max_brightness : 0)) < 0)
        return ret;

    return 0;
}

int led_close(led_t *led) {
    if (led->fd < 0)
        return 0;

    /* Close fd */
    if (close(led->fd) < 0)
        return _led_error(led, LED_ERROR_CLOSE, errno, "Closing LED 'brightness'");

    led->fd = -1;

    return 0;
}

void led_free(led_t *led) {
    free(led);
}

int led_get_brightness(led_t *led, unsigned int *brightness) {
    char buf[16];
    int ret;

    /* Read fd */
    if ((ret = read(led->fd, buf, sizeof(buf))) < 0)
        return _led_error(led, LED_ERROR_IO, errno, "Reading LED 'brightness'");

    /* Rewind */
    if (lseek(led->fd, 0, SEEK_SET) < 0)
        return _led_error(led, LED_ERROR_IO, errno, "Rewinding LED 'brightness'");

    /* Null-terminate over newline */
    buf[ret] = '\0';

    *brightness = strtoul(buf, NULL, 10);

    return 0;
}

int led_get_max_brightness(led_t *led, unsigned int *max_brightness) {
    char led_path[P_PATH_MAX];
    char buf[16];
    int fd, ret;

    /* Read max_brightness */
    snprintf(led_path, sizeof(led_path), "/sys/class/leds/%s/max_brightness", led->name);
    if ((fd = open(led_path, O_RDONLY)) < 0)
        return _led_error(led, LED_ERROR_QUERY, errno, "Opening LED 'max_brightness'");
    if ((ret = read(fd, buf, sizeof(buf))) < 0) {
        int errsv = errno;
        close(fd);
        return _led_error(led, LED_ERROR_QUERY, errsv, "Reading LED 'max_brightness'");
    }
    if (close(fd) < 0)
        return _led_error(led, LED_ERROR_QUERY, errno, "Closing LED 'max_brightness'");

    /* Null-terminate over newline */
    buf[ret] = '\0';

    *max_brightness = strtoul(buf, NULL, 10);

    return 0;
}

int led_set_brightness(led_t *led, unsigned int brightness) {
    char buf[16];
    int len;

    if (brightness > led->max_brightness)
        return _led_error(led, LED_ERROR_ARG, 0, "Brightness out of bounds (max is %u)", led->max_brightness);

    len = snprintf(buf, sizeof(buf), "%u\n", brightness);

    /* Write fd */
    if (write(led->fd, buf, len) < 0)
        return _led_error(led, LED_ERROR_IO, errno, "Writing LED 'brightness'");

    /* Rewind */
    if (lseek(led->fd, 0, SEEK_SET) < 0)
        return _led_error(led, LED_ERROR_IO, errno, "Rewinding LED 'brightness'");

    return 0;
}

int led_fd(led_t *led) {
    return led->fd;
}

int led_name(led_t *led, char *str, size_t len) {
    strncpy(str, led->name, len);
    return 0;
}

int led_tostring(led_t *led, char *str, size_t len) {
    unsigned int brightness;
    char brightness_str[16];
    unsigned int max_brightness;
    char max_brightness_str[16];

    if (led_get_brightness(led, &brightness) < 0)
        strncpy(brightness_str, "<error>", sizeof(brightness_str));
    else
        snprintf(brightness_str, sizeof(brightness_str), "%u", brightness);

    if (led_get_max_brightness(led, &max_brightness) < 0)
        strncpy(max_brightness_str, "<error>", sizeof(max_brightness_str));
    else
        snprintf(max_brightness_str, sizeof(max_brightness_str), "%u", max_brightness);

    return snprintf(str, len, "LED %s (brightness=%s, max_brightness=%s)", led->name, brightness_str, max_brightness_str);
}

int led_errno(led_t *led) {
    return led->error.c_errno;
}

const char *led_errmsg(led_t *led) {
    return led->error.errmsg;
}
